### PR TITLE
Fixing missing custom metrics

### PR DESCRIPTION
**Pull request by mmontgomery-square**

### CODE REVIEW

File: pkg/telemetry/closeable_dogstatsd_sink.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+func (s *CloseableDogStatsdSink) AddSample(key []string, val float32) {
+	s.AddSampleWithLabels(key, val, nil)
+}
+
+func (s *CloseableDogStatsdSink) SetGaugeWithLabels(key []string, val float32, labels []metrics.Label) {
+	flatKey, tags := s.getFlatkeyAndCombinedLabels(key, labels)
+	rate := 1.0
+	_ = s.client.Gauge(flatKey, float64(val), tags, rate)
+}
```
Comment (mmontgomery-square) [inline pkg/telemetry/closeable_dogstatsd_sink.go:104]: The linter was angry that we didn't explicitly ignore the error from s.client.Gauge (and other methods) so now we are.

File: pkg/telemetry/closeable_metric_sink_adapter.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package telemetry
+
+import "github.com/armon/go-metrics"
+
+type CloseableMetricSinkAdapter struct {
+	delegate metrics.MetricSink
+}
+
+var _ CloseableMetricSink = (*CloseableMetricSinkAdapter)(nil)
```
Comment (mmontgomery-square) [inline pkg/telemetry/closeable_metric_sink_adapter.go:9]: This struct just allows us to take a plain ol' MetricSink and slap a Closeable() method onto it, in effect. In practice it's used to adapt the built-in BlackholeSink which gets used when no telemetry socket is specified.

File: pkg/telemetry/telemetry.go [hunk 3]
```diff
@@ -18,10 +17,13 @@ const (
 )
 
 type Telemetry struct {
+	sink    CloseableMetricSink
 	Metrics *metrics.Metrics
 	labels  []metrics.Label
 }
 
+var _ Closeable = (*Telemetry)(nil)
```
Comment (mmontgomery-square) [inline pkg/telemetry/telemetry.go:25]: In order to support consistent emission of metrics we now keep a reference to the metric sink (which must be closeable) and expose a Close() method on Telemetry which then closes the underlying sink.

File: cmd/spiffe-aws-assume-role/cli/cli.go [hunk 4]
```diff
@@ -38,21 +38,25 @@ type CliContext struct {
 	Telemetry         *telemetry.Telemetry
 }
 
-func (c *CredentialsCmd) Run(context *CliContext) error {
+func (c *CredentialsCmd) Run(context *CliContext) (err error) {
 	c.configureLogger(context.Logger)
 
-	telemetry, err := c.configureTelemetry()
+	t, err := c.configureTelemetry()
```
[the remark this answers is not in the document]
Comment (isemaya-square) [reply cmd/spiffe-aws-assume-role/cli/cli.go:44]: Makes sense - if it's not used that much anyway, seems reasonable. I thought maybe we could use `telemetryConfigured`, but that seems excessively wordy 😅

File: pkg/telemetry/closeable.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package telemetry
+
+type Closeable interface {
+	Close()
+}
```
[the remark this answers is not in the document]
Comment (isemaya-square) [reply pkg/telemetry/closeable.go:5]: Eh, not sure if it's a huge deal either way.

**Review by isemaya-square — APPROVED** [review]

File: pkg/telemetry/closeable_dogstatsd_sink.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		client:            client,
+		hostName:          hostName,
+		propagateHostname: false,
+	}
+	return sink, nil
+}
+
+func (s *CloseableDogStatsdSink) Close() {
+	_ = s.client.Close()
+}
```
[the remark this answers is not in the document]
Comment (mbyczkowski) [reply pkg/telemetry/closeable_dogstatsd_sink.go:36]: You might be able to just embed the `statsd.Client` -- see https://golang.org/doc/effective_go#embedding

[the remark this answers is not in the document]
Comment (mmontgomery-square) [reply pkg/telemetry/closeable_dogstatsd_sink.go:36]: The problem is that embedding doesn't change the visibility of the field that I want to access. However, upon further reflection (ha!) I've decided to embed the original struct and then have the Close() method on the outer struct just use reflection to get at the statsd client field.

File: pkg/telemetry/closeable_dogstatsd_sink.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	return sink, nil
+}
+
+func (s *CloseableDogStatsdSink) Close() {
+	structType := reflect.ValueOf(s.DogStatsdSink).Elem()
+	privateField := structType.FieldByName("client")
+	fieldValue := reflect.NewAt(privateField.Type(), unsafe.Pointer(privateField.UnsafeAddr())).Elem()
+	statsdClient := fieldValue.Interface().(*statsd.Client)
+	_ = statsdClient.Close()
+}
```
Comment (mmontgomery-square) [inline pkg/telemetry/closeable_dogstatsd_sink.go:35]: I'd love to say I figured this out on my own, but no:

https://stackoverflow.com/questions/42664837/how-to-access-unexported-struct-fields/43918797#43918797

**Review by mbyczkowski — APPROVED** [review]
There's some good context in a bunch of your comments that could be useful as code comments IMHO

**Pull request merged**